Split out a base class for PathGraph.

RoosterDragon opened the pull request:
Move the domain logic involved into a base class named `DensePathGraph`. The base class contains all the domain logic necessary to traverse a graph including concepts such as custom movement layer.

`PathGraph` becomes responsible for proving a backing array for the pathfinding information, and is where the pooling logic lives instead, helping split the two concepts out.

----

This split is motivated by #19591, in which we will want to create a second derived class that covers only a portion of the map area. Moving all the domain logic into a base class gears us up for that nicely.

Reviewer aid: GitHub doesn't detect copies in its diff window, so here's a diff with copy detection enabled. This makes finding the differences in `DensePathGraph.cs` much clearer rather than it showing up as a wholly new file.

<details>
  <summary>DensePathGraph.cs Diff</summary>

```diff
diff --git a/OpenRA.Mods.Common/Pathfinder/PathGraph.cs b/OpenRA.Mods.Common/Pathfinder/DensePathGraph.cs
similarity index 74%
copy from OpenRA.Mods.Common/Pathfinder/PathGraph.cs
copy to OpenRA.Mods.Common/Pathfinder/DensePathGraph.cs
index c6f4aa883e..7272f7bf4a 100644
--- a/OpenRA.Mods.Common/Pathfinder/PathGraph.cs
+++ b/OpenRA.Mods.Common/Pathfinder/DensePathGraph.cs
@@ -1,6 +1,6 @@
-#region Copyright & License Information
+﻿#region Copyright & License Information
 /*
- * Copyright 2007-2021 The OpenRA Developers (see AUTHORS)
+ * Copyright 2007-2022 The OpenRA Developers (see AUTHORS)
  * This file is part of OpenRA, which is free software. It is made
  * available to you under the terms of the GNU General Public License
  * as published by the Free Software Foundation, either version 3 of
@@ -17,16 +17,15 @@
 namespace OpenRA.Mods.Common.Pathfinder
 {
 	/// <summary>
-	/// A dense pathfinding graph that supports a search over all cells within a map.
-	/// It implements the ability to cost and get connections for cells, and supports <see cref="ICustomMovementLayer"/>.
+	/// A dense pathfinding graph that implements the ability to cost and get connections for cells,
+	/// and supports <see cref="ICustomMovementLayer"/>. Allows searching over a dense grid of cells.
+	/// Derived classes are required to provide backing storage for the pathfinding information.
 	/// </summary>
-	sealed class PathGraph : IPathGraph
+	abstract class DensePathGraph : IPathGraph
 	{
-		public const int PathCostForInvalidPath = int.MaxValue;
-		public const short MovementCostForUnreachableCell = short.MaxValue;
 		const int LaneBiasCost = 1;
 
-		readonly ICustomMovementLayer[] customMovementLayers;
+		protected readonly ICustomMovementLayer[] CustomMovementLayers;
 		readonly int customMovementLayersEnabledForLocomotor;
 		readonly Locomotor locomotor;
 		readonly Actor actor;
@@ -37,14 +36,12 @@ sealed class PathGraph : IPathGraph
 		readonly bool laneBias;
 		readonly bool inReverse;
 		readonly bool checkTerrainHeight;
-		readonly CellInfoLayerPool.PooledCellInfoLayer pooledLayer;
-		readonly CellLayer<CellInfo>[] cellInfoForLayer;
 
-		public PathGraph(CellInfoLayerPool layerPool, Locomotor locomotor, Actor actor, World world, BlockedByActor check,
+		protected DensePathGraph(Locomotor locomotor, Actor actor, World world, BlockedByActor check,
 			Func<CPos, int> customCost, Actor ignoreActor, bool laneBias, bool inReverse)
 		{
-			customMovementLayers = world.GetCustomMovementLayers();
-			customMovementLayersEnabledForLocomotor = customMovementLayers.Count(cml => cml != null && cml.EnabledForLocomotor(locomotor.Info));
+			CustomMovementLayers = world.GetCustomMovementLayers();
+			customMovementLayersEnabledForLocomotor = CustomMovementLayers.Count(cml => cml != null && cml.EnabledForLocomotor(locomotor.Info));
 			this.locomotor = locomotor;
 			this.world = world;
 			this.actor = actor;
@@ -54,17 +51,18 @@ sealed class PathGraph : IPathGraph
 			this.laneBias = laneBias;
 			this.inReverse = inReverse;
 			checkTerrainHeight = world.Map.Grid.MaximumTerrainHeight > 0;
+		}
+
+		public abstract CellInfo this[CPos node] { get; set; }
 
-			// As we support a search over the whole map area,
-			// use the pool to grab the CellInfos we need to track the graph state.
-			// This allows us to avoid the cost of allocating large arrays constantly.
-			// PERF: Avoid LINQ
-			pooledLayer = layerPool.Get();
-			cellInfoForLayer = new CellLayer<CellInfo>[customMovementLayers.Length];
-			cellInfoForLayer[0] = pooledLayer.GetLayer();
-			foreach (var cml in customMovementLayers)
-				if (cml != null && cml.EnabledForLocomotor(locomotor.Info))
-					cellInfoForLayer[cml.Index] = pooledLayer.GetLayer();
+		/// <summary>
+		/// Determines if a candidate neighbouring position is
+		/// allowable to be returned in a <see cref="GraphConnection"/>.
+		/// </summary>
+		/// <param name="neighbor">The candidate cell. This might not lie within map bounds.</param>
+		protected virtual bool NeighborAllowable(CPos neighbor)
+		{
+			return true;
 		}
 
 		// Sets of neighbors for each incoming direction. These exclude the neighbors which are guaranteed
@@ -127,23 +125,28 @@ public List<GraphConnection> GetConnections(CPos position)
 			{
 				var dir = directions[i];
 				var neighbor = position + dir;
+				if (!NeighborAllowable(neighbor))
+					continue;
 
 				var pathCost = GetPathCostToNode(position, neighbor, dir);
-				if (pathCost != PathCostForInvalidPath &&
+				if (pathCost != PathGraph.PathCostForInvalidPath &&
 					this[neighbor].Status != CellStatus.Closed)
 					validNeighbors.Add(new GraphConnection(neighbor, pathCost));
 			}
 
 			if (layer == 0)
 			{
-				foreach (var cml in customMovementLayers)
+				foreach (var cml in CustomMovementLayers)
 				{
 					if (cml == null || !cml.EnabledForLocomotor(locomotor.Info))
 						continue;
 
 					var layerPosition = new CPos(position.X, position.Y, cml.Index);
+					if (!NeighborAllowable(layerPosition))
+						continue;
+
 					var entryCost = cml.EntryMovementCost(locomotor.Info, layerPosition);
-					if (entryCost != MovementCostForUnreachableCell &&
+					if (entryCost != PathGraph.MovementCostForUnreachableCell &&
 						CanEnterNode(position, layerPosition) &&
 						this[layerPosition].Status != CellStatus.Closed)
 						validNeighbors.Add(new GraphConnection(layerPosition, entryCost));
@@ -152,11 +155,14 @@ public List<GraphConnection> GetConnections(CPos position)
 			else
 			{
 				var groundPosition = new CPos(position.X, position.Y, 0);
-				var exitCost = customMovementLayers[layer].ExitMovementCost(locomotor.Info, groundPosition);
-				if (exitCost != MovementCostForUnreachableCell &&
-					CanEnterNode(position, groundPosition) &&
-					this[groundPosition].Status != CellStatus.Closed)
-					validNeighbors.Add(new GraphConnection(groundPosition, exitCost));
+				if (NeighborAllowable(groundPosition))
+				{
+					var exitCost = CustomMovementLayers[layer].ExitMovementCost(locomotor.Info, groundPosition);
+					if (exitCost != PathGraph.MovementCostForUnreachableCell &&
+						CanEnterNode(position, groundPosition) &&
+						this[groundPosition].Status != CellStatus.Closed)
+						validNeighbors.Add(new GraphConnection(groundPosition, exitCost));
+				}
 			}
 
 			return validNeighbors;
@@ -166,16 +172,16 @@ bool CanEnterNode(CPos srcNode, CPos destNode)
 		{
 			return
 				locomotor.MovementCostToEnterCell(actor, srcNode, destNode, check, ignoreActor)
-				!= MovementCostForUnreachableCell;
+				!= PathGraph.MovementCostForUnreachableCell;
 		}
 
 		int GetPathCostToNode(CPos srcNode, CPos destNode, CVec direction)
 		{
 			var movementCost = locomotor.MovementCostToEnterCell(actor, srcNode, destNode, check, ignoreActor);
-			if (movementCost != MovementCostForUnreachableCell)
+			if (movementCost != PathGraph.MovementCostForUnreachableCell)
 				return CalculateCellPathCost(destNode, direction, movementCost);
 
-			return PathCostForInvalidPath;
+			return PathGraph.PathCostForInvalidPath;
 		}
 
 		int CalculateCellPathCost(CPos neighborCPos, CVec direction, short movementCost)
@@ -187,8 +193,8 @@ int CalculateCellPathCost(CPos neighborCPos, CVec direction, short movementCost)
 			if (customCost != null)
 			{
 				var customCellCost = customCost(neighborCPos);
-				if (customCellCost == PathCostForInvalidPath)
-					return PathCostForInvalidPath;
+				if (customCellCost == PathGraph.PathCostForInvalidPath)
+					return PathGraph.PathCostForInvalidPath;
 
 				cellCost += customCellCost;
 			}
@@ -209,15 +215,11 @@ int CalculateCellPathCost(CPos neighborCPos, CVec direction, short movementCost)
 			return cellCost;
 		}
 
-		public CellInfo this[CPos pos]
-		{
-			get => cellInfoForLayer[pos.Layer][pos];
-			set => cellInfoForLayer[pos.Layer][pos] = value;
-		}
+		protected virtual void Dispose(bool disposing) { }
 
 		public void Dispose()
 		{
-			pooledLayer.Dispose();
+			Dispose(true);
 		}
 	}
 }
```
</details>

RoosterDragon committed:
Split out a base class for PathGraph.

Move the domain logic involved into a base class named DensePathGraph. The base class contains all the domain logic necessary to traverse a graph including concepts such as custom movement layer.

PathGraph becomes responsible for proving a backing array for the pathfinding information, and is where the pooling logic lives instead, helping split the two concepts out.

pchote commented:
It feels odd IMO for `PathGraph` to have a more general name than `DensePathGraph` despite inheriting from and being more specific than it. Can we rename that to something like `CellPathGraph`?

RoosterDragon commented:
All the graphs involve cells, so that's not very distinctive. Something like `MapPathGraph` might work - as this one specifically covers the whole map.

It might be worth moving `PathCostForInvalidPath` and `MovementCostForUnreachableCell` to the `IPathGraph` interface instead.

RoosterDragon committed:
Rename PathGraph to MapPathGraph.

Move PathCostForInvalidPath and MovementCostForUnreachableCell constants into a new static class, PathGraph.

RoosterDragon commented:
Have renamed to `MapPathGraph`. Can't move the constants to the interface, so leaving there for now.

pchote commented:
If those constants were exposed as a property instead, is the JIT smart enough to inline their access?

Otherwise, would it make sense for the constants to be defined on DensePathGraph or a small static `PathGraph` class?

RoosterDragon commented:
The last option sounds good, have amended the commit accordingly.

pchote reviewed (dismissed):
Code changes LGTM but not tested yet.

abcdefg30 commented on OpenRA.Mods.Common/Pathfinder/DensePathGraph.cs:
```diff
@@ -209,15 +215,13 @@ int CalculateCellPathCost(CPos neighborCPos, CVec direction, short movementCost)
 		}
 
-		public CellInfo this[CPos pos]
+		protected virtual void Dispose(bool disposing)
```
```suggestion
		protected virtual void Dispose(bool disposing) { }
```

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/bcc3089842609bfa30fe85bc3cbb2b873463a28e...254a70992d2d8b1c739711e2cca28b1d0f68a30a)
